indexmanager/index_manager.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 The Rode Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexmanager
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v7"
+	"go.uber.org/zap"
+)
+
+//go:generate counterfeiter -generate
+
+//counterfeiter:generate -o ../mocks . IndexManager
+type IndexManager interface {
+	MappingsRegistry
+	IndexRepository
+	MigrationOrchestrator
+	// Initialize loads document kind mappings from the path specified in Config.MappingsPath.
+	// Then, using the prefix from Config.IndexPrefix, it finds any indices associated with the application; and, if
+	// necessary, runs a migration to apply schema changes.
+	Initialize(context.Context) error
+}
+
+type indexManager struct {
+	MappingsRegistry
+	IndexRepository
+	MigrationOrchestrator
+}
+
+func NewIndexManager(logger *zap.Logger, client *elasticsearch.Client, config *Config) IndexManager {
```

Comment:
I think I like this approach better than the traditional "do all setup in `main.go`", which I suppose is even more important when this is meant to be consumed as a library.  I guess the only trade off is that this code won't be covered by unit tests, but I think that's worth it.

Reply:
As another approach, I played around with [`fx`](https://github.com/uber-go/fx) a while ago and liked that quite a bit. It composes dependencies at runtime, so there's a hit to startup, but it worked pretty well.

Reply:
Nice, I haven't tried that yet.  the only other candidate we've tried in the past is [wire](https://github.com/google/wire), but we weren't a fan of that.  I'd be interested to see what `fx` looks like though.